napari/util/keybindings.py

```diff
@@ -0,0 +1,250 @@
+            pass
+        return
+    elif isinstance(func, str):
+        new_seq = normalize_key_sequence(func)
+        _rebind_key(seq, new_seq, keybindings)
+
+    _bind_key(seq, func, keybindings)
+
+
+def rebind_key(orig_seq, new_seq, keybindings):
+    orig_seq = normalize_key_sequence(orig_seq)
+    new_seq = normalize_key_sequence(new_seq)
+
+    _rebind_key(orig_seq, new_seq, keybindings)
+
+
+def _bind_keys_validate_pair(pair):
+    if not (isinstance(pair, Sequence) and not isinstance(pair, STRINGTYPES)):
+        raise TypeError(f'expected {pair} to be a sequence, got {type(pair)}')
+    key, op = pair
+    if not isinstance(key, str):
+        raise TypeError(f'key {key} must be a string, not {type(key)}')
+    if op is not None and not isinstance(op, (str, Callable)):
+        raise TypeError(f'operation must be a string, function or None')
+
+
+def _bind_keys_is_valid_pair(pair):
+    try:
+        _bind_keys_validate_pair(pair)
+    except TypeError:
+        return False
+    return True
+
+
+def bind_keys_normalize_input(args, kwargs):
+    if len(args) == 1:
+        # bind_keys((('asdf', func), ('sdf', None)))
+        # is equivalent to
+        # bind_keys(('asdf', func), ('sdf', None))
+        args = args[0]
+
+        if isinstance(args, Mapping):
+            args = args.items()  # dict -> (N, 2) view
+
+        if isinstance(args, Iterable):
+            if isinstance(args, STRINGTYPES):
+                raise TypeError()
+            args = tuple(args)
+
+    if len(args) == 2 and _bind_keys_is_valid_pair(args):
+        # bind_keys('asdf', func) == bind_keys(('asdf', func))
+        args = (args,)
+
+    args += tuple(kwargs.items())
+
+    return args
+
+
+def bind_keys(key_pairs, keybindings):
+    for pair in key_pairs:
+        _bind_keys_validate_pair(pair)
+        bind_key(pair[0], pair[1], keybindings)
+
+
+def _bind_keys_method(self, *args, **kwargs):
+    # name it 'self' so it looks less conspicuous to inspection tools
+    keybindings = self
+    bind_keys(bind_keys_normalize_input(args, kwargs),
+              keybindings)
+
+
+class bind_keys_method:
+    def __init__(self, *, keybindings, class_keybindings=None):
+        self.keybindings = keybindings
+        self.class_keybindings = class_keybindings
+
+    def __get__(self, obj, objtype):
+        if obj is None:  # used on class
+            if self.class_keybindings is None:
+                return _bind_keys_method
+            keybindings = getattr(objtype, self.class_keybindings)
+        else:
+            keybindings = getattr(obj, self.keybindings)
+
+        return types.MethodType(_bind_keys_method, keybindings)
+
+
+def _bind_key_method(self, seq, func=UNDEFINED):
+    keybindings = self
+
+    if func is UNDEFINED:  # used as a decorator
+        def inner(func):
+            bind_key(seq, func, keybindings)
+            return func
+
+        return inner
+
+    bind_key(seq, func, keybindings)
+
+
+class bind_key_method:
```

Comment:
All this logic is super complicated. I know you like metaprogramming but you should add clear and extensive comments that tell the rest of us mortals what exactly this is doing. :stuck_out_tongue_winking_eye: A docstring with a tiny example would be super useful.